File: 47.全排列-ii.go
/*
 * @lc app=leetcode.cn id=47 lang=golang
 *
 * [47] 全排列 II
 */
package main

import "sort"

// @lc code=start
func permuteUnique(nums []int) [][]int {
	sort.Ints(nums)
	var res [][]int
	n := len(nums)
	used := make([]bool, n)
	var path []int
	var depth int
	var permuteUniqueDfs func()
	permuteUniqueDfs = func() {
		if depth == n {
			res = append(res, append([]int{}, path...))
			return
		}
		for i := 0; i < len(nums); i++ {
			if used[i] {
				continue
			}
			if i > 0 && nums[i] == nums[i-1] && !used[i-1] {
				continue
			}
			path = append(path, nums[i])
			used[i] = true
			depth++
			permuteUniqueDfs()
			depth--
			used[i] = false
			path = path[:len(path)-1]
		}
	}
	permuteUniqueDfs()
	return res
}

// @lc code=end

File: 47.全排列-ii.rs
/*
 * @lc app=leetcode.cn id=47 lang=rust
 *
 * [47] 全排列 II
 */

// @lc code=start
impl Solution {
    pub fn permute_unique(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut res = vec![];
        let len = nums.len();
        if len == 0 {
            return res;
        }
        let mut used = vec![false; len];
        let mut path = vec![];
        let mut nums = nums;
        nums.sort_unstable();
        Solution::dfs(&nums, &mut path, &mut used, &mut res);
        res
    }

    pub fn dfs(nums: &Vec<i32>, path: &mut Vec<i32>, used: &mut Vec<bool>, res: &mut Vec<Vec<i32>>) {
        if path.len() == nums.len() {
            res.push(path.clone());
            return;
        }
        for i in 0..nums.len() {
            if used[i] {
                continue;
            }
            if i > 0 && nums[i] == nums[i - 1] && !used[i - 1] {
                continue;
            }
            path.push(nums[i]);
            used[i] = true;
            Solution::dfs(&nums, path, used, res);
            used[i] = false;
            path.pop();
        }
    }
}
// @lc code=end

